server/gfmd/user.c

```diff
@@ -419,6 +355,154 @@ user_enter_auth_id(
 	return (GFARM_ERR_NO_ERROR);
 }
 
+static gfarm_error_t
+user_auth_id_modify_internal(struct user *u,
+		char *auth_id_type, char *auth_user_id,
+		bool *need_to_update_dbp, bool *need_to_addp)
+{
+	enum auth_user_id_type auth_user_id_type = 0;
+	char *old_auth_user_id = NULL;
+	struct gfarm_hash_entry *entry;
+	gfarm_error_t e;
+
+	*need_to_update_dbp = false;
+
+	if ((e = gfarm_auth_user_id_type_from_name(
+		(char *)auth_id_type,
+		&auth_user_id_type)) != GFARM_ERR_NO_ERROR) {
+		return (e);
+	}
+
+	old_auth_user_id = u->auth_user_id[auth_user_id_type];
+
+	if (old_auth_user_id != NULL) {
+		if (strcmp(u->auth_user_id[auth_user_id_type],
+				auth_user_id) == 0) {
+			return (GFARM_ERR_NO_ERROR);
+		}
+
+		struct user_auth_key key = {
+			auth_user_id_type,
+			u->auth_user_id[auth_user_id_type]
+		};
+
+		entry = gfarm_hash_lookup(user_id_map, &key, sizeof(key));
```

Comment:
This is not a good place to call gfarm_hash_purge().
Please think about a failure of strdup_log() below.
If strdup_log() fails, this hash table becomes incosistent against u->auth_user_id[auth_user_id_type];
Thus, it's better to call gfarm_hash_purge() at the end of this function.
Also, this gfarm_hash_lookup() must always succes, thus calling gfarm_hash_lookup() is unnecessary, just call gfarm_hash_purge, and if it returns 0 (i.e. the key not found), call gflog_fatal(, "unexpected inconsistency user %s, auth_id_type %s, old auth_user_id %s", ...)